ruby/ql/lib/codeql/ruby/frameworks/ActionDispatch.qll

```diff
@@ -0,0 +1,889 @@
+   */
+  private class ResourcesRouteBlock extends NestedRouteBlock, TResourcesRouteBlock {
+    private MethodCall call;
+    private Block block;
+
+    ResourcesRouteBlock() { this = TResourcesRouteBlock(parent, call, block) }
+
+    override string getAPrimaryQlClass() { result = "ResourcesRouteBlock" }
+
+    override Stmt getAStmt() { result = block.getAStmt() }
+
+    /**
+     * Gets the `resources` call that gives rise to this route block.
+     */
+    MethodCall getDefiningMethodCall() { result = call }
+
+    override string getPathComponent() {
+      exists(string resource | call.getArgument(0).getConstantValue().isStringOrSymbol(resource) |
+        result = resource + "/:" + singularize(resource) + "_id"
+      )
+    }
+
+    override string getControllerComponent() { result = "" }
+
+    override string toString() { result = call.toString() }
+
+    override Location getLocation() { result = call.getLocation() }
+  }
+
+  /**
+   * A route block that is guarded by a conditional statement.
+   * For example:
+   * ```rb
+   * if Rails.env.test?
+   *   get "/foo/bar", to: "foo#bar"
+   * end
+   * ```
+   * We ignore the condition and analyze both branches to obtain as
+   * much routing information as possible.
+   */
+  private class ConditionalRouteBlock extends NestedRouteBlock, TConditionalRouteBlock {
+    private ConditionalExpr e;
+
+    ConditionalRouteBlock() { this = TConditionalRouteBlock(parent, e) }
+
+    override string getAPrimaryQlClass() { result = "ConditionalRouteBlock" }
+
+    override Stmt getAStmt() { result = e.getBranch(_).(StmtSequence).getAStmt() }
+
+    override string getPathComponent() { none() }
+
+    override string getControllerComponent() { none() }
+
+    override string toString() { result = e.toString() }
+
+    override Location getLocation() { result = e.getLocation() }
+  }
+
+  /**
+   * A route block defined by a call to `namespace`.
+   * ```rb
+   * namespace :admin do
+   *   resources :posts
+   * end
+   * ```
+   * https://api.rubyonrails.org/classes/ActionDispatch/Routing/Mapper/Scoping.html#method-i-namespace
+   */
+  private class NamespaceRouteBlock extends NestedRouteBlock, TNamespaceRouteBlock {
+    private MethodCall call;
+    private Block block;
+
+    NamespaceRouteBlock() { this = TNamespaceRouteBlock(parent, call, block) }
+
+    override Stmt getAStmt() { result = block.getAStmt() }
+
+    override string getPathComponent() { result = this.getNamespace() }
+
+    override string getControllerComponent() { result = this.getNamespace() }
+
+    private string getNamespace() {
+      call.getArgument(0).getConstantValue().isStringOrSymbol(result)
+    }
+
+    override string toString() { result = call.toString() }
+
+    override Location getLocation() { result = call.getLocation() }
+  }
+
+  /**
+   * A route configuration. See `Route` for more info
+   */
+  newtype TRoute =
```

Comment:
private?

Reply:
It probably makes sense to cache this and `TRouteBlock` (in a private cached module to make sure they are evaluated in the same stage).